File: packages/py-build/scripts/verify-groups.mjs
#!/usr/bin/env node
// Host-side gate over the freshly linked grouped side modules
// (dist/groups/<bundle>.so), runnable BEFORE any packaging touches them:
//
//   node scripts/verify-groups.mjs               # verify every census bundle
//   node scripts/verify-groups.mjs --spike mpl-deps --pkgs kiwisolver
//                                                # verify dist/groups/spike-<b>.so
//                                                # against only the built pkgs'
//                                                # extensions + structural compare
//                                                # vs the upstream wheel .so(s)
//
// Per group .so: dylink.0 present + needed == [] (nothing vendored), PyInit
// export set == the census union for the bundle (exact equality — a missing
// PyInit means a dropped extension, an extra one means a stray input), and
// closed-world vs the CURRENT main module: every census-relevant import
// (env/GOT.mem/GOT.func, invoke_*/plumbing excluded) resolves from
// mainExports ∪ own exports ∪ the js-glue allowlist ({exit} today). The
// grouped imports ⊆ old per-extension union, so this holds against the
// EXISTING main before link.rsp regenerates — the de-risking property the
// packaging swap leans on.
import { existsSync, readFileSync } from 'node:fs';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { inflateRawSync } from 'node:zlib';
import { censusImports, parseWasm } from './lib/wasm-parse.mjs';

const pkgRoot = resolve(dirname(fileURLToPath(import.meta.url)), '..');
const groupsDir = join(pkgRoot, 'dist/groups');
const groups = JSON.parse(readFileSync(join(pkgRoot, 'config/dso-groups/groups.json'), 'utf8'));
const config = JSON.parse(readFileSync(join(pkgRoot, 'build.config.json'), 'utf8'));
const mainWasmPath = [join(pkgRoot, '../../web/public/py-dev/fork/pyodide.asm.wasm'), join(pkgRoot, 'dist/raw/pyodide.asm.wasm')].find(
  existsSync,
);
if (!mainWasmPath) throw new Error('no pyodide.asm.wasm found (staged public or dist/raw)');
const mainExports = new Set(parseWasm(readFileSync(mainWasmPath), false).exports);
const JS_GLUE_ALLOW = new Set(['exit']);

const args = process.argv.slice(2);
const opt = (name) => {
  const i = args.indexOf(name);
  return i >= 0 ? args[i + 1] : null;
};
const spikeBundle = opt('--spike');
const spikePkgs = opt('--pkgs') ? new Set(opt('--pkgs').split(',')) : null;

// Minimal zip central-directory reader for wheels (fetch-wheels' walk).
function* zipEntries(buf) {
  let eocd = buf.length - 22;
  while (buf.readUInt32LE(eocd) !== 0x06054b50) eocd--;
  let off = buf.readUInt32LE(eocd + 16);
  const count = buf.readUInt16LE(eocd + 10);
  for (let i = 0; i < count; i++) {
    if (buf.readUInt32LE(off) !== 0x02014b50) throw new Error('bad central directory');
    const method = buf.readUInt16LE(off + 10);
    const csize = buf.readUInt32LE(off + 20);
    const nameLen = buf.readUInt16LE(off + 28);
    const extraLen = buf.readUInt16LE(off + 30);
    const cmtLen = buf.readUInt16LE(off + 32);
    const lho = buf.readUInt32LE(off + 42);
    const name = buf.toString('utf8', off + 46, off + 46 + nameLen);
    yield {
      name,
      read() {
        const lhNameLen = buf.readUInt16LE(lho + 26);
        const lhExtraLen = buf.readUInt16LE(lho + 28);
        const dataOff = lho + 30 + lhNameLen + lhExtraLen;
        const raw = buf.subarray(dataOff, dataOff + csize);
        return method === 8 ? inflateRawSync(raw) : Buffer.from(raw);
      },
    };
    off += 46 + nameLen + extraLen + cmtLen;
  }
}

const failures = [];
const verifyOne = (bundle, exts, soPath) => {
  if (!existsSync(soPath)) {
    failures.push(`${bundle}: ${soPath} missing — run the recipes loop`);
    return null;
  }
  const w = parseWasm(readFileSync(soPath));
  const name = soPath.split('/').pop();
  if (!w.dylink) failures.push(`${name}: no dylink.0 section`);
  else if (w.dylink.needed.length) failures.push(`${name}: NEEDED not empty: ${w.dylink.needed.join(', ')}`);
  const wantPyinits = new Set(exts.map((e) => e.pyinit));
  const gotPyinits = new Set(w.exports.filter((x) => x.startsWith('PyInit')));
  for (const p of wantPyinits) if (!gotPyinits.has(p)) failures.push(`${name}: census PyInit missing from exports: ${p}`);
  for (const p of gotPyinits) if (!wantPyinits.has(p)) failures.push(`${name}: stray PyInit export not in census: ${p}`);
  const exportSet = new Set(w.exports);
  const unresolved = new Set();
  for (const { field } of censusImports(w)) {
    if (!mainExports.has(field) && !exportSet.has(field) && !JS_GLUE_ALLOW.has(field)) unresolved.add(field);
  }
  if (unresolved.size)
    failures.push(
      `${name}: ${unresolved.size} imports outside main ∪ self ∪ glue: ${[...unresolved].slice(0, 8).join(', ')}${unresolved.size > 8 ? ' …' : ''}`,
    );
  const imp = censusImports(w);
  console.log(
    `${name}: ${readFileSync(soPath).length.toLocaleString()} B · ${w.exports.length} exports (${gotPyinits.size} PyInit) · ` +
      `${imp.length} census imports · dylink mem ${w.dylink?.memSize ?? '??'} tbl ${w.dylink?.tableSize ?? '??'} · needed [${w.dylink?.needed ?? '?'}]`,
  );
  return w;
};

if (spikeBundle) {
  const g = groups.bundles[spikeBundle];
  if (!g) throw new Error(`unknown bundle ${spikeBundle}`);
  const exts = g.extensions.filter((e) => !spikePkgs || spikePkgs.has(e.pkg));
  const w = verifyOne(spikeBundle, exts, join(groupsDir, `spike-${spikeBundle}.so`));
  if (w) {
    // Structural compare vs the upstream wheel .so(s) — census-level, NOT
    // byte-equality (container paths leak into debug strings by design).
    for (const e of exts) {
      const pin = config.recipes.wheels[e.pkg];
      const wheel = join(pkgRoot, '.cache/wheels', pin.file);
      if (!existsSync(wheel)) {
        console.log(`(no ${pin.file} in .cache/wheels — skipping upstream compare for ${e.dottedName})`);
        continue;
      }
      let up = null;
      for (const entry of zipEntries(readFileSync(wheel))) {
        if (entry.name === e.wheelSoPath) up = parseWasm(entry.read());
      }
      if (!up) throw new Error(`${e.wheelSoPath} not in ${pin.file}`);
      const upImports = new Set(censusImports(up).map((i) => i.field));
      const groupExports = new Set(w.exports);
      const escaped = [...new Set(censusImports(w).map((i) => i.field))].filter((f) => !upImports.has(f) && !groupExports.has(f));
      console.log(
        `vs upstream ${e.dottedName}: upstream ${up.exports.length} exports / ${censusImports(up).length} census imports · ` +
          `dylink mem ${up.dylink?.memSize} tbl ${up.dylink?.tableSize} · group-not-in-upstream imports: ${escaped.length ? escaped.join(', ') : '(none)'}`,
      );
      if (escaped.length) failures.push(`spike ${e.dottedName}: group imports ${escaped.length} symbols upstream never imported`);
    }
  }
} else {
  for (const [bundle, g] of Object.entries(groups.bundles)) {
    verifyOne(bundle, g.extensions, join(groupsDir, `${bundle}.so`));
  }
}

if (failures.length) {
  console.error(`\nverify-groups FAILURES (${failures.length}):`);
  for (const f of failures) console.error(`  ${f}`);
  process.exit(1);
}
console.log('\nverify-groups OK');

File: packages/py-build/src/avlo_py_build/verify_groups.py
"""Host-side gate over the freshly linked grouped side modules
(dist/groups/<bundle>.so), runnable BEFORE any packaging touches them:

  avlo-build verify-groups                          # every census bundle
  avlo-build verify-groups --spike mpl-deps --pkgs kiwisolver
                                                    # verify dist/groups/
                                                    # spike-<b>.so against only
                                                    # the built pkgs' extensions
                                                    # + structural compare vs
                                                    # the upstream wheel .so(s)

Per group .so: dylink.0 present + needed == [] (nothing vendored), PyInit
export set == the census union for the bundle (exact equality — a missing
PyInit means a dropped extension, an extra one means a stray input), and
closed-world vs the CURRENT main module: every census-relevant import
(env/GOT.mem/GOT.func, invoke_*/plumbing excluded) resolves from
mainExports ∪ own exports ∪ the js-glue allowlist ({exit} today). The
grouped imports ⊆ old per-extension union, so this holds against the
EXISTING main before link.rsp regenerates — the de-risking property the
packaging swap leans on.
"""

import json
import sys
import zipfile

from .config import load
from .paths import FORK_PUBLIC, GROUPS_DIR, PKG_ROOT, RAW_DIR, WHEEL_CACHE
from .wasmmeta import ParsedWasm, census_imports, parse_wasm

JS_GLUE_ALLOW = frozenset(["exit"])


def run(args) -> int:
    groups = json.loads((PKG_ROOT / "config/dso-groups/groups.json").read_text())
    cfg = load()
    main_wasm_path = next(
        (p for p in (FORK_PUBLIC / "pyodide.asm.wasm", RAW_DIR / "pyodide.asm.wasm") if p.exists()), None
    )
    if main_wasm_path is None:
        sys.exit("no pyodide.asm.wasm found (staged public or dist/raw)")
    main_exports = set(parse_wasm(main_wasm_path.read_bytes(), want_dylink=False).exports)

    failures: list[str] = []

    def verify_one(bundle: str, exts: list[dict], so_path) -> ParsedWasm | None:
        if not so_path.exists():
            failures.append(f"{bundle}: {so_path} missing — run the recipes loop")
            return None
        data = so_path.read_bytes()
        w = parse_wasm(data)
        name = so_path.name
        if w.dylink is None:
            failures.append(f"{name}: no dylink.0 section")
        elif w.dylink.needed:
            failures.append(f"{name}: NEEDED not empty: {', '.join(w.dylink.needed)}")
        want_pyinits = {e["pyinit"] for e in exts}
        got_pyinits = {x for x in w.exports if x.startswith("PyInit")}
        for p in sorted(want_pyinits - got_pyinits):
            failures.append(f"{name}: census PyInit missing from exports: {p}")
        for p in sorted(got_pyinits - want_pyinits):
            failures.append(f"{name}: stray PyInit export not in census: {p}")
        export_set = set(w.exports)
        imp = census_imports(w)
        unresolved = list(
            dict.fromkeys(
                field
                for _m, field, _k in imp
                if field not in main_exports and field not in export_set and field not in JS_GLUE_ALLOW
            )
        )
        if unresolved:
            failures.append(
                f"{name}: {len(unresolved)} imports outside main ∪ self ∪ glue: "
                f"{', '.join(unresolved[:8])}{' …' if len(unresolved) > 8 else ''}"
            )
        dl = w.dylink
        print(
            f"{name}: {len(data):,} B · {len(w.exports)} exports ({len(got_pyinits)} PyInit) · "
            f"{len(imp)} census imports · dylink mem {dl.memSize if dl else '??'} tbl {dl.tableSize if dl else '??'} · "
            f"needed [{', '.join(dl.needed) if dl else '?'}]"
        )
        return w

    if args.spike:
        g = groups["bundles"].get(args.spike)
        if g is None:
            sys.exit(f"unknown bundle {args.spike}")
        spike_pkgs = set(args.pkgs.split(",")) if args.pkgs else None
        exts = [e for e in g["extensions"] if spike_pkgs is None or e["pkg"] in spike_pkgs]
        w = verify_one(args.spike, exts, GROUPS_DIR / f"spike-{args.spike}.so")
        if w is not None:
            # Structural compare vs the upstream wheel .so(s) — census-level,
            # NOT byte-equality (container paths leak into debug strings by
            # design).
            for e in exts:
                pin = cfg.recipes.wheels[e["pkg"]]
                wheel = WHEEL_CACHE / pin.file
                if not wheel.exists():
                    print(f"(no {pin.file} in .cache/wheels — skipping upstream compare for {e['dottedName']})")
                    continue
                with zipfile.ZipFile(wheel) as z:
                    try:
                        up = parse_wasm(z.read(e["wheelSoPath"]))
                    except KeyError:
                        sys.exit(f"{e['wheelSoPath']} not in {pin.file}")
                up_imports = {f for _m, f, _k in census_imports(up)}
                group_exports = set(w.exports)
                escaped = [
                    f
                    for f in dict.fromkeys(f for _m, f, _k in census_imports(w))
                    if f not in up_imports and f not in group_exports
                ]
                udl = up.dylink
                print(
                    f"vs upstream {e['dottedName']}: upstream {len(up.exports)} exports / "
                    f"{len(census_imports(up))} census imports · dylink mem {udl.memSize if udl else None} "
                    f"tbl {udl.tableSize if udl else None} · "
                    f"group-not-in-upstream imports: {', '.join(escaped) if escaped else '(none)'}"
                )
                if escaped:
                    failures.append(f"spike {e['dottedName']}: group imports {len(escaped)} symbols upstream never imported")
    else:
        for bundle, g in groups["bundles"].items():
            verify_one(bundle, g["extensions"], GROUPS_DIR / f"{bundle}.so")

    if failures:
        print(f"\nverify-groups FAILURES ({len(failures)}):", file=sys.stderr)
        for f in failures:
            print(f"  {f}", file=sys.stderr)
        return 1
    print("\nverify-groups OK")
    return 0
